arc.m

```diff
@@ -767,39 +773,47 @@ static BOOL setObjectHasWeakRefs(id obj)
 OBJC_PUBLIC id objc_storeWeak(id *addr, id obj)
 {
 	LOCK_FOR_SCOPE(&weakRefLock);
-	WeakRef *oldRef;
-	id old;
-	loadWeakPointer(addr, &old, &oldRef);
-	// If the old and new values are the same, then we don't need to do anything.
-	if (old == obj)
+	BOOL assignNoAndUnregister = NO;
+	if(obj == nil)
 	{
-		return obj;
+		assignNoAndUnregister = YES;
 	}
 	BOOL isGlobalObject = setObjectHasWeakRefs(obj);
-	// If we old ref exists, decrement its reference count.  This may also
-	// delete the weak reference control block.
-	if (oldRef != NULL)
```

Comment:
Now that this is moved later, you are *not* releasing the `WeakRef` structure if assigning a global object to the weak object.  This may mask your bug, because we're now leaking `WeakRef`s on some code paths.